File: bindings/csharp/Umicp.Core/Types/OperationType.cs
using System.Text.Json.Serialization;

namespace Umicp.Core.Types;

/// <summary>
/// Message operation types for UMICP protocol
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum OperationType
{
    /// <summary>
    /// Control message type
    /// </summary>
    Control,

    /// <summary>
    /// Data transfer message type
    /// </summary>
    Data,

    /// <summary>
    /// Acknowledgment message type
    /// </summary>
    Ack,

    /// <summary>
    /// Error message type
    /// </summary>
    Error,

    /// <summary>
    /// Request message type
    /// </summary>
    Request,

    /// <summary>
    /// Response message type
    /// </summary>
    Response
}

File: bindings/kotlin/src/main/kotlin/com/hivellm/umicp/types/OperationType.kt
package com.hivellm.umicp.types

import kotlinx.serialization.Serializable

/**
 * UMICP Operation Types
 *
 * Defines the type of operation being performed in a message envelope.
 */
@Serializable
enum class OperationType {
    /** Control messages for protocol management */
    CONTROL,

    /** Data transfer messages */
    DATA,

    /** Acknowledgment messages */
    ACK,

    /** Error messages */
    ERROR,

    /** Request messages */
    REQUEST,

    /** Response messages */
    RESPONSE;

    companion object {
        /**
         * Convert from string value
         */
        fun fromString(value: String): OperationType? =
            entries.find { it.name.equals(value, ignoreCase = true) }
    }
}

